### PR TITLE
Add rain and fog simulation features

### PR DESCRIPTION
**Pull request by PiotrMrozik**
This code introduces public components for rain and fog simulations. Each simulation can be controlled with dedicated manager (images below).
image
image

### CODE REVIEW

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLNativeAPI.cs [hunk 1]
```diff
@@ -182,6 +182,17 @@ public static extern int rgl_node_points_simulate_snow(ref IntPtr node, float mi
...
+        [DllImport("RobotecGPULidar")]
+        public static extern int rgl_node_points_simulate_rain(ref IntPtr node, float min_range, float max_range, float rain_rate,
+            Int32 num_channels, float beam_divergence , Int32 numerical_threshold, float occupancy_threshold);
+
+        [DllImport("RobotecGPULidar")]
+        public static extern int rgl_node_points_simulate_rain_configure_defaults(IntPtr node, int droplets_id, float full_beam_intensity,
+            float droplets_laser_retro);
+
+        [DllImport("RobotecGPULidar")]
+        public static extern int rgl_node_points_simulate_fog(ref IntPtr node, float attenuationCoefficient, float r1, float r2);
```
Comment (PawelLiberadzki) [inline Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLNativeAPI.cs:194]: `rgl_node_points_simulate_rain_configure_defaults` call is not imported - please add it like is is done for snow and rain. Note that there will be an additional property - FogId - in FogManager.

File: Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public class LidarFogManager : MonoBehaviour
+    {
+        // Singleton pattern
+        public static LidarFogManager Instance { get; private set; }
+
+        // Delegate to notify that the snow model has been changed
+        public delegate void OnNewConfigDelegate();
+        public OnNewConfigDelegate OnNewConfig;
+
+        [field: SerializeField]
```
Comment (PawelLiberadzki) [inline Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs:32]: Please also add header attributes (i.a. [field: Header("Base Settings")]) like it is done in snow and rain managers.

File: Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        [field: Tooltip("Near cross-point of transmitter and receiver beams of Lidar device in meters. Model assumes that lidar is working in bistatic beam configuration.")]
+        [field: Range(0.0f, 1.0f)]
+        public float nearCrossPoint { get; private set; } = 0.01f;
+
+        [field: SerializeField]
+        [field: Tooltip("far cross-point of transmitter and receiver beams of Lidar device. Model assumes that lidar is working in bistatic beam configuration.")]
+        [field: Range(1.0f, 10.0f)]
+        public float farCrossPoint { get; private set; } = 2.0f;
```
Comment (PawelLiberadzki) [inline Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs:49]:
```suggestion
        [field: SerializeField]
        [field: Tooltip("Near cross-point of transmitter and receiver beams of Lidar device in meters. Model assumes that lidar is working in bistatic beam configuration.")]
        [field: Range(0.0f, 1.0f)]
        public float NearCrossPoint { get; private set; } = 0.01f;

        [field: SerializeField]
        [field: Tooltip("Far cross-point of transmitter and receiver beams of Lidar device. Model assumes that lidar is working in bistatic beam configuration.")]
        [field: Range(1.0f, 10.0f)]
        public float FarCrossPoint { get; private set; } = 2.0f;
```
Uppercase letter for properties.

File: Assets/RGLUnityPlugin/Scripts/LidarSensor.cs [hunk 4]
```diff
@@ -246,6 +264,44 @@ private void ApplyConfiguration(BaseLidarConfiguration newConfig)
...
+            // Fog model updates
+            if (rglGraphLidar.HasNode(fogNodeId))
+            {
+                // Update fog parameters only if feature is enabled (expensive operation)
+                if (LidarFogManager.Instance.IsFogEnabled)
+                {
+                    rglGraphLidar.UpdateNodePointsSimulateFog(fogNodeId,
+                        LidarFogManager.Instance.AttenuationCoefficient,
+                        LidarFogManager.Instance.nearCrossPoint,
+                        LidarFogManager.Instance.farCrossPoint);
```
Comment (PawelLiberadzki) [inline Assets/RGLUnityPlugin/Scripts/LidarSensor.cs:299]: Whe you add a call for configuring fog simulation defaults, please also remember to add `.UpdateNodePointsRainDefaults` call below this line.

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLNativeAPI.cs [hunk 5]
```diff
@@ -652,6 +663,24 @@ public static void NodePointsSimulateSnowConfigureDefaults(IntPtr node, int snow
...
+                occupancyThreshold));
+        }
+
+        public static void NodePointsSimulateRainConfigureDefaults(IntPtr node, int dropletsId, float fullBeamIntensity,
+            float dropletsLaserRetro)
+        {
+            CheckErr(rgl_node_points_simulate_rain_configure_defaults(node, dropletsId, fullBeamIntensity, dropletsLaserRetro));
+        }
+
+        public static   void NodePointsSimulateFog(ref IntPtr node, float attenuationCoefficient, float r1, float r2)
```
Comment (PawelLiberadzki) [inline Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLNativeAPI.cs:679]:
```suggestion
        public static void NodePointsSimulateFog(ref IntPtr node, float attenuationCoefficient, float r1, float r2)
```

File: Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs (hunk 3 above)
Comment (PawelLiberadzki) [reply Assets/RGLUnityPlugin/Scripts/LidarFogManager.cs:49]: Please also add information about units of all fog parameters.

File: Assets/AWSIM/Prefabs/RGLWeatherManager.prefab
Comment (PawelLiberadzki) [inline Assets/AWSIM/Prefabs/RGLWeatherManager.prefab:1]: Please adjust specific weather manager parameters in this prefab - e.g. for RainManager parameters start from too heavy rain and I would adjust them to default values from RainManager component. Same for other manager probably.

**Review by PawelLiberadzki — COMMENTED** [review]
Additional note from runtime testing. Except from this, everything seems to work fine in runtime.

**Review by PawelLiberadzki — APPROVED** [review]
LGTM!

**Review by mackierx111 — APPROVED** [review]
I will check this feature in the internal repository.

**Pull request merged**